src/lsp-server/src/backend.rs

```diff
@@ -0,0 +1,254 @@
+            .await;
+
+        self.parse(TextDocumentItem {
+            uri: params.text_document.uri,
+            text: params.content_changes[0].text.clone(),
+            version: params.text_document.version,
+        })
+        .await
+    }
+
+    async fn did_save(&self, _: DidSaveTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+    }
+
+    async fn did_close(&self, _: DidCloseTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file closed!")
+            .await;
+    }
+
+    async fn completion(&self, params: CompletionParams) -> Result<Option<CompletionResponse>> {
+        let _uri = params.text_document_position.text_document.uri;
+        let _position = params.text_document_position.position;
+
+        // TODO: Re enable when position is correct.
+        // Ok(completions.map(CompletionResponse::Array))
+        Ok(None)
+    }
+
+    async fn code_lens(&self, _params: CodeLensParams) -> Result<Option<Vec<CodeLens>>> {
+        let _lenses: Vec<CodeLens> = vec![CodeLens {
+            range: Range {
+                start: Position::new(0, 0),
+                end: Position::new(0, 0),
+            },
+            command: Some(Command {
+                title: "Run".to_string(),
+                command: "materialize.run".to_string(),
+                arguments: None,
+            }),
+            data: None,
+        }];
+
+        // TODO: Re enable when position is correct.
+        // Ok(Some(lenses))
+        Ok(None)
+    }
+}
+
+struct TextDocumentItem {
+    uri: Url,
+    text: String,
+    version: i32,
+}
+
+impl Backend {
+    async fn parse(&self, params: TextDocumentItem) {
+        self.client
+            .log_message(MessageType::INFO, format!("on_change {:?}", params.uri))
+            .await;
+        let rope = ropey::Rope::from_str(&params.text);
+
+        // Parse the text
+        let mz_sql_parser = mz_sql_parser::parser::parse_statements(&params.text);
+
+        match mz_sql_parser {
+            // The parser will return Ok when everything is well written.
+            Ok(results) => {
+                self.client
+                    .log_message(MessageType::INFO, format!("Results: {:?}", results))
+                    .await;
+
+                // Clear the diagnostics in case there were issues before.
+                self.client
+                    .publish_diagnostics(params.uri.clone(), vec![], Some(params.version))
+                    .await;
+            }
+
+            // If there is at least one error the parser will return Err.
+            Err(err_parsing) => {
+                self.client
+                    .log_message(MessageType::INFO, format!("Error: {:?}", err_parsing))
+                    .await;
+
+                let error_position = err_parsing.error.pos;
+                let start = offset_to_position(error_position, &rope).unwrap();
```

Comment:
Should we instead default to `0` (via `unwrap_or_default()`) when we cannot parse a position from the offset? I always get annoyed when LSP servers crash.

Reply:
Ah, just realized this is a `Position`, not `u32`. Might be worth seeing if Position defaults to `(0,0)`, and, if not, `unwrap_or_else(|| Position::new(0, 0))`

Reply:
I agree. Having zero `unwrap()` or `expect()` should be aimed. I made it return `None` to differentiate from the case where is in fact `(0,0)`, and let the function logic decide if it should go to `(0,0)`.